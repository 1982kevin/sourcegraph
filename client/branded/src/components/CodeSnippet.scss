.code-snippet {
    &__container {
        background-color: var(--code-bg);
    }
}
