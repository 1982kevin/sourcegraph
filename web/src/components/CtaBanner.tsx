import * as React from 'react'
import classNames from 'classnames'

interface Props {
    className?: string
    icon: string | React.ReactNode
}

export const CtaBanner = React.memo<Props>(({ icon, className }) => (
    <div className={classNames('web-content private-code-cta shadow d-flex flex-row card', className)}>
        <div className="private-code-cta__icon-column d-flex flex-column align-items-center">
            {typeof icon === 'string' ? <img src={icon} className="private-code-cta__logo" /> : { icon }}
        </div>
        <div>
            <h3>Search your private code</h3>
            <p>
                Set up a self-hosted Sourcegraph instance to search your private repositories on GitHub, GitLab,
                Bitbucket and local installations of Git, Perforce, Subversion and other code repositories.
            </p>
            <a
                href="https://docs.sourcegraph.com/"
                target="_blank"
                rel="noreferrer"
                className="btn btn-primary ga-cta-install-now"
            >
                Install now
            </a>
        </div>
    </div>
))
