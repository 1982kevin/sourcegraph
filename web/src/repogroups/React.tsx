import { RepogroupMetadata } from './types'
import { SearchPatternType } from '../../../shared/src/graphql/schema'
import * as React from 'react'
export const react: RepogroupMetadata = {
    title: 'React',
    name: 'react',
    url: '/react',
    description:
        "Explore React repositories on GitHub. The search examples show how to use React Hook useState with various data types.",
    examples: [
        {
            title: 'Find imports of useState with regex search',
            exampleQuery: <>import [^;]+useState[^;]+ from 'react'</>,
            rawQuery: "import [^;]+useState[^;]+ from 'react'",
            patternType: SearchPatternType.regexp,
        },
        {
            title: 'Examples of useState with an object as the input parameter',
            description: `useState takes a single argument and can accept a primitive, an array or an object.
            Syntax for usage is [state_value, function_to_update_state_value] = useState (initial_state_value).`,
            exampleQuery: <>{'useState({:[string]})'}</>,
            rawQuery: 'useState({:[string]}) count:1000',
            patternType: SearchPatternType.structural,
        },
        {
            title: 'Examples of useState with any JavaScript data type as the input parameter',
            exampleQuery: <>useState(:[string])</>,
            rawQuery: 'useState(:[string]) count:1000',
            patternType: SearchPatternType.structural,
        },
        {
            title: 'Examples of useState with any JavaScript data type as the input parameter in only TypeScript files',
            exampleQuery: (
                <>
                    useState(:[string])
                    <span className="search-keyword">lang:</span>typescript
                </>
            ),
            rawQuery: 'useState(:[string]) count:1000 lang:typescript',
            patternType: SearchPatternType.structural,
        },
    ],
    homepageDescription: 'Learn React with code examples.',
    homepageIcon: 'https://cdn4.iconfinder.com/data/icons/logos-3/600/React.js_logo-512.png',
}
