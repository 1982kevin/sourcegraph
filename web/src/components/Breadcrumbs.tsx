import React, { useState, useCallback, useEffect } from 'react'
import ChevronRightIcon from 'mdi-react/ChevronRightIcon'

export interface Breadcrumb {
    key: string
    element: React.ReactNode | null
}

export interface BreadcrumbsProps {
    breadcrumbs: Breadcrumb[]
}

export interface UpdateBreadcrumbsProps {
    setBreadcrumb: (key: string, element: React.ReactNode) => () => void
}

export const useBreadcrumbs = (): BreadcrumbsProps & UpdateBreadcrumbsProps => {
    const [breadcrumbs, setBreadcrumbs] = useState<Breadcrumb[]>([])
    const setBreadcrumb = useCallback((key: string, element: React.ReactNode) => {
        console.log('setBreadcrumb', key, element)
        setBreadcrumbs(breadcrumbs => {
            const index = breadcrumbs.findIndex(breadcrumb => breadcrumb.key === key)
            if (index === -1) {
                return [...breadcrumbs, { key, element }]
            }
            return [...breadcrumbs.slice(0, index), { key, element }, ...breadcrumbs.slice(index + 1)]
        })
        return () => {
            // Replace with null (but remember order in case the key gets set again)
            setBreadcrumb(key, null)
        }
    }, [])
    useEffect(() => console.log(breadcrumbs), [breadcrumbs])
    return {
        breadcrumbs,
        setBreadcrumb,
    }
}

export const Breadcrumbs: React.FunctionComponent<BreadcrumbsProps> = ({ breadcrumbs }) => (
    <>
        {breadcrumbs
            .filter(({ element }) => element !== null)
            .map(({ element, key }, index) => (
                <React.Fragment key={key}>
                    {index !== 0 && <ChevronRightIcon />} {element}
                </React.Fragment>
            ))}
    </>
)
