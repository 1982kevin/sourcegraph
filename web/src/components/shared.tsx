import classNames from 'classnames'
import React from 'react'
import { ActionsNavItems, ActionsNavItemsProps } from '../../../shared/src/actions/ActionsNavItems'
import { CommandListPopoverButton, CommandListPopoverButtonProps } from '../../../shared/src/commandPalette/CommandList'
import {
    EditorCompletionWidget,
    EditorCompletionWidgetProps,
} from '../../../shared/src/components/completion/EditorCompletionWidget'
import { HoverOverlay, HoverOverlayProps } from '../../../shared/src/hover/HoverOverlay'

// Components from shared with web-styling class names applied

interface WebHoverOverlayState {}
export class WebHoverOverlay extends React.PureComponent<HoverOverlayProps, WebHoverOverlayState> {
    constructor(props: HoverOverlayProps) {
        super(props)
    }

    public render(): JSX.Element | null {
        console.log(this.alertStorageKey('hi'))
        const filteredAlerts = (this.props.hoverOrError?.alerts || []).filter((alert) => localStorage.getItem(this.alertStorageKey(alert.dismissalType)) !== 'true')
        console.log(this.props.hoverOrError?.alerts || [], filteredAlerts)
        const props = this.props.hoverOrError?.alerts ? {...this.props, hoverOrError: {...this.props.hoverOrError, alerts: filteredAlerts}} : this.props

        return <HoverOverlay
            {...props}
            className="card"
            iconClassName="icon-inline"
            iconButtonClassName="btn btn-icon"
            actionItemClassName="btn btn-secondary"
            infoAlertClassName="alert alert-info"
            errorAlertClassName="alert alert-danger"
            onAlertDismissed={this.onAlertDismissedCallback()}
        />
    }

    private onAlertDismissedCallback(): (alertType: string) => void {
        const thisClass = this;
        return alertType => {
            localStorage.setItem(this.alertStorageKey(alertType), 'true')
            thisClass.forceUpdate()
        }
    }

    private alertStorageKey(alertType: string): string {
        return `WebHoverOverlay/alert/${alertType}/dismissed`
    }
}
WebHoverOverlay.displayName = 'WebHoverOverlay'

export const WebCommandListPopoverButton: React.FunctionComponent<CommandListPopoverButtonProps> = props => (
    <CommandListPopoverButton
        {...props}
        buttonClassName="btn btn-link"
        popoverClassName="popover"
        popoverInnerClassName="border rounded overflow-hidden"
        formClassName="form"
        inputClassName="form-control px-2 py-1 rounded-0"
        listClassName="list-group list-group-flush list-unstyled"
        actionItemClassName="list-group-item list-group-item-action px-2"
        selectedActionItemClassName="active border-primary"
        noResultsClassName="list-group-item text-muted"
    />
)
WebCommandListPopoverButton.displayName = 'WebCommandListPopoverButton'

export const WebActionsNavItems: React.FunctionComponent<ActionsNavItemsProps> = ({
    listClass,
    listItemClass,
    actionItemClass,
    actionItemIconClass,
    ...props
}) => (
        <ActionsNavItems
            {...props}
            listClass={classNames(listClass, 'nav')}
            listItemClass={classNames(listItemClass, 'nav-item')}
            actionItemClass={classNames(actionItemClass, 'nav-link')}
            actionItemIconClass={classNames(actionItemIconClass, 'icon-inline')}
        />
    )
WebActionsNavItems.displayName = 'WebActionsNavItems'

const completionWidgetListItemClassName = 'completion-widget-dropdown__item d-flex align-items-center p-2'

export const WebEditorCompletionWidget: React.FunctionComponent<EditorCompletionWidgetProps> = props => (
    <EditorCompletionWidget
        {...props}
        listClassName="completion-widget-dropdown d-block list-unstyled rounded p-0 m-0 mt-3"
        listItemClassName={completionWidgetListItemClassName}
        selectedListItemClassName="completion-widget-dropdown__item--selected bg-primary"
        loadingClassName={completionWidgetListItemClassName}
        noResultsClassName={completionWidgetListItemClassName}
    />
)
WebEditorCompletionWidget.displayName = 'WebEditorCompletionWidget'
