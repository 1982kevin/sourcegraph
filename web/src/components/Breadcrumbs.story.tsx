import { storiesOf } from '@storybook/react'
import React from 'react'
import { Breadcrumbs } from './Breadcrumbs'
import webStyles from '../SourcegraphWebApp.scss'

const { add } = storiesOf('web/Breadcrumbs', module).addDecorator(story => (
    <>
        <style>{webStyles}</style>
        <div className="layout__app-router-container">
            <div className="container mt-3">{story()}</div>
        </div>
    </>
))

add('Empty breadcrumbs', () => <Breadcrumbs breadcrumbs={[]} />)

add(
    'Three levels deep',
    () => (
        <Breadcrumbs
            breadcrumbs={[
                <span className="">Level 2</span>,
                <span className="">Level 3</span>,
                <span className="">Page name</span>,
            ]}
        />
    ),
    {
        design: {
            type: 'figma',
            url: 'https://www.figma.com/file/BkY8Ak997QauG0Iu2EqArv/Sourcegraph-Components?node-id=230%3A83',
        },
    }
)
