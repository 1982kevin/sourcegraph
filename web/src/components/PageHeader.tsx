import React from 'react'
import { Breadcrumbs } from './Breadcrumbs'

interface Props {
    title: string
    icon: React.ReactNode
    breadcrumbs: React.ReactNode[]
    actions?: React.ReactNode
    badge?: string // TODO: consider support for multiple badges
}

export const PageHeader = ({ title, icon, actions, badge, breadcrumbs }: Props) => (
    <>
        <Breadcrumbs breadcrumbs={breadcrumbs} />
        <div className="d-flex align-items-center">
            <h1 className="flex-grow-1 text-nowrap">
                {icon} {title}
                {badge && (
                    <sup>
                        <span className="badge badge-primary text-uppercase">{badge}</span>
                    </sup>
                )}
            </h1>
            {actions}
        </div>
    </>
)
