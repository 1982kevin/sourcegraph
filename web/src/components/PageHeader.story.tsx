import { storiesOf } from '@storybook/react'
import React from 'react'
import webStyles from '../SourcegraphWebApp.scss'
import { PageHeader } from './PageHeader'
import PuzzleOutlineIcon from 'mdi-react/PuzzleOutlineIcon'
import { text } from '@storybook/addon-knobs'
import ClockOutlineIcon from 'mdi-react/ClockOutlineIcon'

const { add } = storiesOf('web/PageHeader', module).addDecorator(story => (
    <>
        <style>{webStyles}</style>
        <div className="layout__app-router-container">
            <div className="container mt-3">{story()}</div>
        </div>
    </>
))

add('Basic', () => <PageHeader title="Basic" breadcrumbs={['Page name']} icon={<PuzzleOutlineIcon />} />)

add(
    'Complex',
    () => {
        const title = text('Title', 'Page name')
        return (
            <PageHeader
                title={title}
                breadcrumbs={['Level 2', 'Level 3', title]}
                icon={<ClockOutlineIcon className="inline-icon" />}
                badge="prototype"
            />
        )
    },
    {
        design: {
            type: 'figma',
            url: 'https://www.figma.com/file/BkY8Ak997QauG0Iu2EqArv/Sourcegraph-Components?node-id=206%3A71',
        },
    }
)
