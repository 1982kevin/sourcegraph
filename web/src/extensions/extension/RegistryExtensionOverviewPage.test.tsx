import { noop } from 'lodash'
import React from 'react'
import { RegistryExtensionOverviewPage } from './RegistryExtensionOverviewPage'
import { PageTitle } from '../../components/PageTitle'
import { createMemoryHistory } from 'history'
import { Router } from 'react-router'
import { mount } from 'enzyme'

jest.mock('mdi-react/GithubIcon', () => 'GithubIcon')

describe('RegistryExtensionOverviewPage', () => {
    afterEach(() => {
        PageTitle.titleSet = false
    })
    test('renders', () => {
        const history = createMemoryHistory()
        expect(
            mount(
                <Router history={history}>
                    <RegistryExtensionOverviewPage
                        eventLogger={{ logViewEvent: noop }}
                        extension={{
                            id: 'x',
                            rawManifest: '{}',
                            manifest: {
                                url: 'https://example.com',
                                activationEvents: ['*'],
                                categories: ['Programming languages', 'Other'],
                                tags: ['T1', 'T2'],
                                readme: '**A**',
                                repository: {
                                    url: 'https://github.com/foo/bar',
                                    type: 'git',
                                },
                            },
                        }}
                        history={history}
                    />
                </Router>
            )
        ).toMatchSnapshot()
    })

    describe('categories', () => {
        test('filters out unrecognized categories', () => {
            const history = createMemoryHistory()
            const output = mount(
                <Router history={history}>
                    <RegistryExtensionOverviewPage
                        eventLogger={{ logViewEvent: noop }}
                        extension={{
                            id: 'x',
                            rawManifest: '',
                            manifest: {
                                url: 'https://example.com',
                                activationEvents: ['*'],
                                categories: ['Programming languages', 'invalid', 'Other'],
                            },
                        }}
                        history={createMemoryHistory()}
                    />
                </Router>
            )
            const foundCategories: string[] = []
            output.find('.list-inline-item').forEach(element => {
                foundCategories.push(element.getDOMNode().textContent!)
            })
            expect(foundCategories).toEqual(['Other', 'Programming languages' /* no 'invalid' */])
        })
    })
})
