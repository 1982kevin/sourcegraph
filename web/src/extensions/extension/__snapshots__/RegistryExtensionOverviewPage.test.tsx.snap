// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistryExtensionOverviewPage renders 1`] = `
<Router
  history="[History]"
>
  <RegistryExtensionOverviewPage
    eventLogger={
      Object {
        "logViewEvent": [Function],
      }
    }
    extension={
      Object {
        "id": "x",
        "manifest": Object {
          "activationEvents": Array [
            "*",
          ],
          "categories": Array [
            "Programming languages",
            "Other",
          ],
          "readme": "**A**",
          "repository": Object {
            "type": "git",
            "url": "https://github.com/foo/bar",
          },
          "tags": Array [
            "T1",
            "T2",
          ],
          "url": "https://example.com",
        },
        "rawManifest": "{}",
      }
    }
    history="[History]"
  >
    <div
      className="registry-extension-overview-page d-flex flex-wrap"
    >
      <PageTitle
        title="x"
      />
      <div
        className="registry-extension-overview-page__readme mr-3"
      >
        <ExtensionReadme
          extension={
            Object {
              "id": "x",
              "manifest": Object {
                "activationEvents": Array [
                  "*",
                ],
                "categories": Array [
                  "Programming languages",
                  "Other",
                ],
                "readme": "**A**",
                "repository": Object {
                  "type": "git",
                  "url": "https://github.com/foo/bar",
                },
                "tags": Array [
                  "T1",
                  "T2",
                ],
                "url": "https://example.com",
              },
              "rawManifest": "{}",
            }
          }
          history="[History]"
        >
          <Markdown
            dangerousInnerHTML="<p><strong>A</strong></p>
"
            history="[History]"
          >
            <div
              className="markdown"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<p><strong>A</strong></p>
",
                }
              }
              onClick={[Function]}
            />
          </Markdown>
        </ExtensionReadme>
      </div>
      <aside
        className="registry-extension-overview-page__sidebar"
      >
        <div
          className="mb-3"
        >
          <h3
            className="mb-0"
          >
            Categories
          </h3>
          <ul
            className="list-inline registry-extension-overview-page__categories"
          >
            <li
              className="list-inline-item mb-2 small"
              key="0"
            >
              <Link
                className="rounded border p-1"
                to="/extensions?query=category%3AOther"
              >
                <LinkAnchor
                  className="rounded border p-1"
                  href="/extensions?query=category%3AOther"
                  navigate={[Function]}
                >
                  <a
                    className="rounded border p-1"
                    href="/extensions?query=category%3AOther"
                    onClick={[Function]}
                  >
                    Other
                  </a>
                </LinkAnchor>
              </Link>
            </li>
            <li
              className="list-inline-item mb-2 small"
              key="1"
            >
              <Link
                className="rounded border p-1"
                to="/extensions?query=category%3A%22Programming+languages%22"
              >
                <LinkAnchor
                  className="rounded border p-1"
                  href="/extensions?query=category%3A%22Programming+languages%22"
                  navigate={[Function]}
                >
                  <a
                    className="rounded border p-1"
                    href="/extensions?query=category%3A%22Programming+languages%22"
                    onClick={[Function]}
                  >
                    Programming languages
                  </a>
                </LinkAnchor>
              </Link>
            </li>
          </ul>
        </div>
        <div
          className="mb-3"
        >
          <h3
            className="mb-0"
          >
            Tags
          </h3>
          <ul
            className="list-inline registry-extension-overview-page__tags"
          >
            <li
              className="list-inline-item mb-2 small"
              key="0"
            >
              <Link
                className="rounded border p-1"
                to="/extensions?query=tag%3AT1"
              >
                <LinkAnchor
                  className="rounded border p-1"
                  href="/extensions?query=tag%3AT1"
                  navigate={[Function]}
                >
                  <a
                    className="rounded border p-1"
                    href="/extensions?query=tag%3AT1"
                    onClick={[Function]}
                  >
                    T1
                  </a>
                </LinkAnchor>
              </Link>
            </li>
            <li
              className="list-inline-item mb-2 small"
              key="1"
            >
              <Link
                className="rounded border p-1"
                to="/extensions?query=tag%3AT2"
              >
                <LinkAnchor
                  className="rounded border p-1"
                  href="/extensions?query=tag%3AT2"
                  navigate={[Function]}
                >
                  <a
                    className="rounded border p-1"
                    href="/extensions?query=tag%3AT2"
                    onClick={[Function]}
                  >
                    T2
                  </a>
                </LinkAnchor>
              </Link>
            </li>
          </ul>
        </div>
        <small
          className="text-muted"
        >
          <dl
            className="border-top pt-2"
          >
            <dt
              className="border-top pt-2"
            >
              Extension ID
            </dt>
            <dd>
              x
            </dd>
            <dt
              className="border-top pt-2"
            >
              Resources
            </dt>
            <dd
              className="border-bottom pb-2"
            >
              <a
                className="d-block"
                href="https://example.com"
                rel="nofollow noopener noreferrer"
                target="_blank"
              >
                Source code (JavaScript)
              </a>
              <div
                className="d-flex"
              >
                <GithubIcon
                  className="icon-inline mr-1"
                />
                <a
                  className="d-block"
                  href="https://github.com/foo/bar"
                  rel="nofollow noreferrer noopener"
                  target="_blank"
                >
                  Repository
                </a>
              </div>
            </dd>
          </dl>
        </small>
      </aside>
    </div>
  </RegistryExtensionOverviewPage>
</Router>
`;
